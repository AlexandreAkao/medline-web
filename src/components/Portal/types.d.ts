interface IPortalProps extends IChildren {
  wrapperId?: string;
}
