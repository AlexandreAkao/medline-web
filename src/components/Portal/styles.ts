import { css } from 'styled-components';

export const portalStyles = css`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
`;
