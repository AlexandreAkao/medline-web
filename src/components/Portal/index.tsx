import { useState, useLayoutEffect, useCallback } from 'react';
import { createPortal } from 'react-dom';

import { portalStyles } from 'components/Portal/styles';

function Portal({ children, wrapperId = 'react-portal-wrapper' }: IPortalProps) {
  const [wrapperElement, setWrapperElement] = useState<HTMLElement>();

  const createWrapperAndAppendToBody = useCallback(() => {
    const createdWrapperElement = document.createElement('div');
    createdWrapperElement.setAttribute('id', wrapperId);
    createdWrapperElement.setAttribute('style', portalStyles.join());
    document.body.appendChild(createdWrapperElement);
    return createdWrapperElement;
  }, [wrapperId]);

  useLayoutEffect(() => {
    let element = document.getElementById(wrapperId);
    let systemCreated = false;

    if (!element) {
      systemCreated = true;
      element = createWrapperAndAppendToBody();
    }
    setWrapperElement(element);

    return () => {
      if (systemCreated && element && element.parentNode) {
        element.parentNode.removeChild(element);
      }
    };
  }, [createWrapperAndAppendToBody, wrapperId]);

  if (!wrapperElement) return null;

  return createPortal(children, wrapperElement);
}

export default Portal;
