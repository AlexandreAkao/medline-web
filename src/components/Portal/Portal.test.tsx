import { render } from '@testing-library/react';

import Portal from 'components/Portal';

describe('Portal', () => {
  const renderComponent = (customProps: Partial<IPortalProps> = {}) => {
    const props: IPortalProps = {
      children: 'Portal example',
      ...customProps,
    };
    return render(<Portal {...props} />);
  };

  it('should match with snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render Portal', () => {
    const { container } = renderComponent();
    expect(container).toBeInTheDocument();
  });
});
