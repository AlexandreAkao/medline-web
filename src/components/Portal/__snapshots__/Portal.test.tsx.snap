// Vitest Snapshot v1

exports[`Portal > should match with snapshot 1`] = `<div />`;
